apis/field_error.go

```diff
@@ -73,10 +73,84 @@ func (fe *FieldError) ViaField(prefix ...string) *FieldError {
 				strings.Join(append(prefix, oldPath), "."))
 		}
 	}
+
+	newPaths = flattenIndexes(newPaths)
+
 	fe.Paths = newPaths
 	return fe
 }
 
+// ViaIndex is used to attach an index to the next via field provided.
+// For example, if a type recursively validates a parameter that has a collection:
+//  for i, c := range spec.Collection {
+//    if err := doValidation(c); err != nil {
+//      return err.ViaIndex(i).ViaField("collection")
+//    }
+//  }
+func (fe *FieldError) ViaIndex(index ...int) *FieldError {
+	if fe == nil {
+		return nil
+	}
+
+	prefix := []string{}
+	for _, i := range index {
+		prefix = append(prefix, fmt.Sprintf("[%d]", i))
+	}
+
+	return fe.ViaField(prefix...)
+}
+
+// ViaKey is used to attach a key to the next via field provided.
+// For example, if a type recursively validates a parameter that has a collection:
+//  for k, v := range spec.Bag. {
+//    if err := doValidation(v); err != nil {
+//      return err.ViaKey(k).ViaField("bag")
+//    }
+//  }
+func (fe *FieldError) ViaKey(key ...string) *FieldError {
+	if fe == nil {
+		return nil
+	}
+
+	prefix := []string{}
+	for _, k := range key {
+		prefix = append(prefix, fmt.Sprintf("[%s]", k))
+	}
+
+	return fe.ViaField(prefix...)
+}
+
+// flattenIndexes takes in a set of paths and looks for chances to flatten
+// objects that have index prefixes, examples:
+//   err([0]).ViaField(bar).ViaField(foo) -> foo.bar.[0] converts to foo.bar[0]
+//   err(bar).ViaIndex(0).ViaField(foo) -> foo.[0].bar converts to foo[0].bar
+//   err(bar).ViaField(foo).ViaIndex(0) -> [0].foo.bar converts to [0].foo.bar
+//   err(bar).ViaIndex(0).ViaIndex[1].ViaField(foo) -> foo.[0].[1].bar converts to foo[1][0].bar
+func flattenIndexes(paths []string) []string {
```

Comment:
Hmm, let's see if we can rework to kill this?

Reply:
I tried to not add this, but the issue is the index has to come in first, and then the field is attached to it so the flattening logic has to be somewhere...